src/Native/System.Native/pal_mount.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_mount.h"
+#include <assert.h>
+#include <string.h>
+#include <errno.h>
+
+// Check if we should use getmntinfo or /proc/mounts
+#if HAVE_MNTINFO
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#include <mntent.h>
+#define STRING_BUFFER_SIZE 8192
+#endif
+
+#if HAVE_MNTINFO
+
+int GetMountInfo(MountPointFound onFound)
+{
+    // getmntinfo returns pointers to OS-internal structs, so we don't need to worry about free'ing the object
+    int result = 0;
+    struct statfs* mounts = nullptr;
+    int count = getmntinfo(&mounts, 0);
+    if (count > 0)
+    {
+        for (int32_t i = 0; i < count; i++)
+        {
+            int32_t nameLength = strlen(mounts[i].f_mntonname);
+            onFound(mounts[i].f_mntonname, nameLength);
+        }
+    }
+    else
+    {
+        // Something went wrong so set our result to errno
+        result = errno;
+    }
+    
+    return result;
+}
+
+#else
+
+int GetMountInfo(MountPointFound onFound)
+{
+    int result = 0;
+    FILE* fp = setmntent("/proc/mounts", MNTOPT_RO);
+    if (fp == nullptr)
+    {
+        result = errno;
+    }
+    else
+    {
+        // The _r version of getmntent needs all buffers to be passed in; however, we don't know how big of a string
+        // buffer we will need, so pick something that seems like it will be big enough.
```

Comment:
Looks like the function succeeds (return non-null), errno stays at 0, and no data is inserted into the buffer, and the char array length is 0; however, the pointer into the /proc/mounts file is preserved so the loop WILL terminate. To handle this case, I'll add a loop that will double the buffer if we fail. This will add a bit of complexity here but if we want this to be rock-solid then it might be worth it